### PR TITLE
Add support of .opensearch-knn-model as system index to transport actions

### PR DESCRIPTION
**Pull request by martin-gaievski**
### Description
We need to stash thread context in transport actions in addition to ModelDao to address all integ tests failures like in log below. This change compliments previous PR https://github.com/opensearch-project/k-NN/pull/827

```
org.opensearch.client.ResponseException: method [POST], host [https://localhost:9200/], URI [/_plugins/_knn/models/test-model-id/_train], status line [HTTP/1.1 403 Forbidden]

    {"error":{"root_cause":[{"type":"security_exception","reason":"no permissions for [] and User [name=admin, backend_roles=[admin], requestedTenant=null]"}],"type":"security_exception","reason":"no permissions for [] and User [name=admin, backend_roles=[admin], requestedTenant=null]"},"status":403}

        at __randomizedtesting.SeedInfo.seed([99F6FF5E9B7971F9:4372F53206C75963]:0)

        at app//org.opensearch.client.RestClient.convertResponse(RestClient.java:375)

        at app//org.opensearch.client.RestClient.performRequest(RestClient.java:345)

        at app//org.opensearch.client.RestClient.performRequest(RestClient.java:320)

        at app//org.opensearch.knn.KNNRestTestCase.trainModel(KNNRestTestCase.java:1099)

        at app//org.opensearch.knn.KNNRestTestCase.ingestDataAndTrainModel(KNNRestTestCase.java:1229)

        at app//org.opensearch.knn.KNNRestTestCase.ingestDataAndTrainModel(KNNRestTestCase.java:1215)

        at app//org.opensearch.knn.plugin.action.RestDeleteModelHandlerIT.testDeleteTrainingModel(RestDeleteModelHandlerIT.java:95)
```
 
### Issues Resolved
https://github.com/opensearch-project/security/issues/2265
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/knn/common/TaskRunner.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+import org.opensearch.client.Client;
+import org.opensearch.common.util.concurrent.ThreadContext;
+
+import java.util.function.Supplier;
+
+/**
+ * Class abstracts execution of runnable or function in specific context
+ */
+public class TaskRunner {
```
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/common/TaskRunner.java:16]: can we rename this class to ThreadContextHelper?

Comment (martin-gaievski) [reply src/main/java/org/opensearch/knn/common/TaskRunner.java:16]: yes, this makes more sense. I could came up with a good name yesterday so took the straightforward one. I'll push a commit soon

File: src/main/java/org/opensearch/knn/indices/ModelDao.java [hunk 2]
```diff
@@ -219,21 +218,14 @@ public void create(ActionListener<CreateIndexResponse> actionListener) throws IO
             if (isCreated()) {
                 return;
             }
-            runWithStashedThreadContext(() -> {
```
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/indices/ModelDao.java:222]: why we need to remove this from runWithStashedThreadContext?

File: src/main/java/org/opensearch/knn/indices/ModelDao.java [hunk 3]
```diff
@@ -303,9 +295,8 @@ private void putInternal(Model model, ActionListener<IndexResponse> listener, Do
                 parameters.put(KNNConstants.MODEL_BLOB_PARAMETER, base64Model);
             }
 
-            final IndexRequestBuilder indexRequestBuilder = ModelDao.runWithStashedThreadContext(
-                () -> client.prepareIndex(MODEL_INDEX_NAME)
-            );
+            IndexRequestBuilder indexRequestBuilder = client.prepareIndex(MODEL_INDEX_NAME);
```
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/indices/ModelDao.java:298]: same as above.

File: src/main/java/org/opensearch/knn/indices/ModelDao.java [hunk 4]
```diff
@@ -369,12 +361,12 @@ private ActionListener<IndexResponse> getUpdateModelMetadataListener(
 
         @SneakyThrows
```
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/indices/ModelDao.java:362]: sneakythrows are not supposed to be used in src files. it is mainly for tests. please remove this.

File: src/main/java/org/opensearch/knn/indices/ModelDao.java (hunk 2 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/knn/indices/ModelDao.java:222]: we do context stashing at the transport action level, which is a caller for this method, so keeping this functionality here is redundant.

File: src/main/java/org/opensearch/knn/indices/ModelDao.java (hunk 3 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/knn/indices/ModelDao.java:298]: we do context stashing at the transport action level, keeping this functionality here is redundant

File: src/main/java/org/opensearch/knn/indices/ModelDao.java (hunk 4 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/knn/indices/ModelDao.java:362]: ack

File: src/main/java/org/opensearch/knn/indices/ModelDao.java (hunk 3 above)
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/indices/ModelDao.java:298]: Thanks this make sense

**Review by navneet1v — APPROVED** [review]

**Review by naveentatikonda — APPROVED** [review]

**Review by navneet1v — APPROVED** [review]

**Pull request merged**